### ISSUE TITLE
Backpack, Rucksack, Bag of Holding, Stash, you name it.

### ISSUE THREAD

**Issue by oddlama** [oddlama/vane#129]
Shulkerboxes are pretty much already bag-extensions. The only thing missing would be re-texturing the item and preventing it from being placed. Step two: Bags can be enchanted(?) to make them greedy, meaning they suck up items that enter your inventory when there's already an item of the same kind in the bag. Thoughts?

(Maybe this could be merged with #126, meaning that e.g. the opening and drop-in functionality only applies to crafted bags and not to regular shulkerboxes.)

idea for crafting recipe:
```
- L -
L S L
- L -

L: Leather
S: Shulkerbox
```

**Comment by ryantheleach** [comment]
There's also minecraft:bundle to consider, but my gut feeling is it's a little early to play with it, being a preview item.

**Comment by oddlama** [comment]
And don't forget that the bundle is mostly useless the way it is now xD
If mojang decides to do anything of the sorts, we can always remove it from vane, or even change items to whatever mojang came up with :D

**Comment by ryantheleach** [comment]
I'd be keen to see Shulkerboxes left alone as-is, and maybe introduce bags as the interesting version that can be opened inside an inventory etc.

Thoughts on a greedy shulker, a greedy shulkerbox placed in the world could potentially act like an area of effect hopper? Not sure about perf implications though, or tracking state of block entities being loaded, bleh.

Enchantments could be a cool way to implant the mechanic onto bags, and fitting with it's somewhat 'magic' nature.

But it furthers the link between trifles and vane-enchantments.

Opening shulker boxes is annoying in protected areas, but I also dislike giving the player new functionality for free, unless it's to mitigate a downside of introduced mechanics to make server play bareble.

Netherite reinforcing a shulkerbox seems a little costly to turn it into a bag, but for an item that requires end city raiding, just combining it with leather seems a little cheap.

I'm wondering if bags could be a middle ground upgrade between bundles and shulkers, to deal with the increasing inventory strain of survival play.

My bag proposal would be Use the bundle recipe as proposed by VanillaTweaks to let players use bundles.

![image](https://user-images.githubusercontent.com/141554/152461589-132688ed-d0b5-4a38-849d-12a656892e7e.png)

or

![image](https://user-images.githubusercontent.com/141554/152461614-89bc6e2c-8a2b-4748-b01a-e31629da2c23.png)

and let the consequences fall where they may regarding future compatibility :S.

Bag's would then be an upgraded form of bundles that can hold more. I don't think it's likely to be possible to just 'let a bundle hold more' so this would likely need to be implemented either in a clever way, or by dumbly retexturing droppers and using the NBT storage of droppers to hold the items.

I'm reluctant to mess around with the 'drop items into bags' functionality too much, even if it's an intended future feature of bundles, just because the inventory related coded is so fragile. (I've encountered remapped controls breaking most item based gui's from most plugins in some form, due to either spigot bugs, or just the way vanilla works / sends packets) and the possible interaction with chest sorting client mods etc.

But anyway, it would be cool to see a relatively expensive mid-game recipe for bags, that didn't rely on shulkers, that had less storage then shulkers, and have shulkers as the 'end game' upgrade for bags.

This would bring inventory management back to a manageable state for players who wish to progress the game at a more leisurely state, rather then the technical / speed runny types.

So TLDR;

My proposed tech tree would be:

Bundles > Bags > ShulkerBag

Leave Shulkers alone by default, but allow them to be crafted into the ShulkerBag, and experiment to see if bundle mechanics can be applied to larger inventory sizes.

Along this vein, we could also consider craftingtableonastick / enderchest without placements.

It makes me wonder whether 'open from an inventory' enchantment could just be universally applicable. e.g. apply it to craftingtables, anvils, grindstones, furnaces?

**Comment by oddlama** [comment]
> Thoughts on a greedy shulker, a greedy shulkerbox placed in the world could potentially act like an area of effect hopper? Not sure about perf implications though, or tracking state of block entities being loaded, bleh.

Possible without performance implications by increasing the item-collection AABB of the internal entity. But I'm also not sure about tracking that state when loading or unloading. Might be very hard to get right for just being a better hopper.

> But it furthers the link between trifles and vane-enchantments.

That can be solved (And it's about time to actually remove that link!). #132

> Netherite reinforcing a shulkerbox seems a little costly to turn it into a bag, but for an item that requires end city raiding, just combining it with leather seems a little cheap.

Definitely. If the bags get all the benefits they need to be more costly.
Be sure to let me know if you have any specific ideas.

> My bag proposal would be Use the bundle recipe as proposed by VanillaTweaks to let players use bundles.

I like giving rabbit hide a use. And it will make player kill cute bunnies to obtain the bags.

> Bag's would then be an upgraded form of bundles that can hold more. I don't think it's likely to be possible to just 'let a bundle hold more' so this would likely need to be implemented either in a clever way, or by dumbly retexturing droppers and using the NBT storage of droppers to hold the items.

Probably the last option.

> I'm reluctant to mess around with the 'drop items into bags' functionality too much, even if it's an intended future feature of bundles, just because the inventory related coded is so fragile. (I've encountered remapped controls breaking most item based gui's from most plugins in some form, due to either spigot bugs, or just the way vanilla works / sends packets) and the possible interaction with chest sorting client mods etc.

If we keep it simple as in rightclick (shouldn't be remapped), I'm sure we can make it work. If anything interferes in the client, it's not a mandatory function and affected players will just have to live without it.

> But anyway, it would be cool to see a relatively expensive mid-game recipe for bags, that didn't rely on shulkers, that had less storage then shulkers, and have shulkers as the 'end game' upgrade for bags.
> [...]
> Leave Shulkers alone by default, but allow them to be crafted into the ShulkerBag, and experiment to see if bundle mechanics can be applied to larger inventory sizes.
> 
> Along this vein, we could also consider craftingtableonastick / enderchest without placements.

Seconded. All of it. The craftingtableonastick seems to be a common theme even :P
Also "Enderbag".

Related note: A friend recently rambled about a variant of an enderchest that has a shared inventory with all other
enderchests of that new variant, but only in the world it is in. Like a shared inventory for the whole world. Not to allow large distance item transfer for yourself, but just to allow all players in the world to share or donate items. Could be problematic on large or open servers with trolls, but cool for private servers.

> It makes me wonder whether 'open from an inventory' enchantment could just be universally applicable. e.g. apply it to craftingtables, anvils, grindstones, furnaces?

I raise to you: anvilonastick, grindstoneonastick furnaceonastick. Only half-kidding xD
I'm slightly opposed to the idea of using enchantments again to make items openable, as from experience I know
that people will put it on everything anyway. So a proper skin and a crafting recipe might be a better option.

**Comment by ryantheleach** [comment]
> Related note: A friend recently rambled about a variant of an enderchest that has a shared inventory with all other
enderchests of that new variant, but only in the world it is in. 

So intentionally making it dimension specific? Could be cool.

It's inevitable that someone is going to want a 'second enderbag' after seeing how the enderchest mod evolved, and with wool colors (and my trolling attempt on one modded server, where I made 4096 of them, using minefactories routers to loot every single one, to prove a point after the admins made a specific rule about stealing from enderchests being free game, that resulted in them deleteing the entire chunks of that base (or i hit a bug))

A potential solution I like, rather then combinations, or private per player, would be to have a crafting recipe entangle them, similar to map cloning.

For sanity's sake, and balance, they would need to be non-interactive with hoppers if placeable in the world.
No matter how much it's requested (It *will* be requested).

**Comment by oddlama** [comment]
> A potential solution I like, rather then combinations, or private per player, would be to have a crafting recipe entangle them, similar to map cloning.

Great idea. Linking via a recipe will also automatically convey that the link is persistent when the chests are relocated. I honestly feel like we need to create a separate module for all these new things that are not quite quality of life improvements but "extensions in the spirit of the community".

> For sanity's sake, and balance, they would need to be non-interactive with hoppers if placeable in the world. No matter how much it's requested (It _will_ be requested).

Definitely. I feel like implementing this would honestly not be a problem at all. I might investigate and make it an option for those that want it, just not enabled by default.

**Comment by ryantheleach** [comment]
I was thinking all the storage related stuff could be grouped together in vane-storage maybe.

**Comment by oddlama** [comment]
Sounds good. That will put us at 10 jar files (9 options)... Not sure if we need to (or even can) do something about that.
`core`, `admin`, `bedtime`, `enchantments`, `permissions`, `portals`, `regions`, `trifles`, `storage`, `additions(?)`...

**Comment by oddlama** [comment]
Implemented Pouch (3x3) and Backpack (3x9) for now.
